### DISCUSSION TITLE
Optimal Route Generation for Multiple Destinations in Python or WASM

### DISCUSSION THREAD

**Discussion by iamshreeram** [organicmaps/organicmaps#8041] [Q&A]
1. Can the OrganicMap data be imported into Python to find the most efficient route between multiple destinations?
2. Is it feasible to utilize the similar logic discussed in the Organic Maps GitHub repository to generate an optimized list of destinations with their respective timings based on a list of coordinates?
3. If the above tasks cannot be accomplished in Python, is it possible to convert the logic into WebAssembly (wasm) for cross-platform usage?

P.S: I don't want to download the entire app. Rather, I prefer to download only the regional data into my python program's memory and utilize the optimal route logic on top of it.

**Comment by biodranik** [comment]
In theory, you can run C++ OM core on top of map data, and use its API from any other language.

**Reply by iamshreeram** [reply]
Thank you for responding, @biodranik . I started out using the Google Maps API and transitioned to OpenStreetMap because of its focus on open data and concerns about censorship.

As a newcomer to C++, diving into OM-Core (~with its **6.5 GB repository size**~) without clear developer-friendly documentation will be quite challenging. Despite this, I have managed to clone the [OM Core](https://github.com/organicmaps/organicmaps) repository and am currently exploring its contents.

My goal is to create a system or module that can generate the most efficient routes for a given set of coordinates or locations, without any risk of censorship now or in the future – a task at which Organic Maps excels.

To achieve this, I am working on the following steps:
1. Obtain a reliable source of uncensored map data for a small region (for instance, `us_texas_dallas.xml` as a hypothetical example).
2. Identify the specific components (certain .cpp files) within the engine that handle route optimization for coordinates from XML or other formats.
3. Modularize the engine using WebAssembly (WASM), or if feasible, create a Python wrapper for the engine that can retrieve necessary region data based on given coordinates and compute the generated path between them.

Once these components are in place, we can develop various applications and interfaces on top of them.

**Reply by iamshreeram** [reply]
Upon further investigation, I discovered that the `pack` is of size approximately 6.4GB. Here is the breakdown of file size contribution. Despite this, the code base is not very large.

```
~/h/p/c/organicmaps ❯❯❯ find . -type f -exec ls -al {} \; | sort -nr -k5 | head -n 25
-r--r--r--  1 home  staff  7001095138 May  1 06:09 ./.git/objects/pack/pack-e46a2eed4b02becca0e7953a68e8d2f80a6493bc.pack
-rw-r--r--  1 home  staff  50994390 May  1 06:09 ./data/dictionary.slf
-rw-r--r--  1 home  staff  47555104 May  1 06:09 ./data/World.mwm
-rw-r--r--  1 home  staff  20461104 May  1 06:09 ./tools/shaders_compiler/linux/mali_compiler_es3/openglessl/libMali-Gxx_r3p0-00rel0.so
-r--r--r--  1 home  staff  20181560 May  1 06:09 ./.git/objects/pack/pack-e46a2eed4b02becca0e7953a68e8d2f80a6493bc.idx
-rw-r--r--  1 home  staff  8581556 May  1 06:09 ./data/WorldCoasts.mwm
-rw-r--r--  1 home  staff  8384030 May  1 06:09 ./data/kml_test_data/test.kmz
-rw-r--r--  1 home  staff  7315682 May  1 06:09 ./3party/opening_hours/opening_hours_integration_tests/opening-count.lst
-rwxr-xr-x  1 home  staff  6498904 May  1 06:09 ./tools/shaders_compiler/macos/glslangValidator
-rwxr-xr-x  1 home  staff  6131664 May  1 06:09 ./tools/shaders_compiler/linux/glslangValidator
-rw-r--r--  1 home  staff  3676511 May  1 06:09 ./data/packed_polygons.bin
-rw-r--r--  1 home  staff  3451900 May  1 06:09 ./data/02_droidsans-fallback.ttf
-rw-r--r--  1 home  staff  3155104 May  1 06:09 ./data/06_code2000.ttf
-rw-r--r--  1 home  staff  3017640 May  1 06:09 ./tools/shaders_compiler/linux/mali_compiler_es3/openglessl/libMali-T600_r13p0-00rel0.so
-rw-r--r--  1 home  staff  2583640 May  1 06:09 ./tools/shaders_compiler/linux/mali_compiler_es3/openglessl/libMali-T600_r12p0-00rel0.so
-r--r--r--  1 home  staff  2396732 May  1 06:09 ./.git/objects/pack/pack-e46a2eed4b02becca0e7953a68e8d2f80a6493bc.rev
-rw-r--r--  1 home  staff  2090971 May  1 06:09 ./data/drules_proto_outdoors_dark.txt
-rw-r--r--  1 home  staff  2088301 May  1 06:09 ./data/drules_proto_outdoors_clear.txt
-rw-r--r--  1 home  staff  2074210 May  1 06:09 ./data/drules_proto.txt
-rw-r--r--  1 home  staff  2012134 May  1 06:09 ./data/drules_proto_dark.txt
-rw-r--r--  1 home  staff  2008831 May  1 06:09 ./data/drules_proto_clear.txt
-rw-r--r--  1 home  staff  1990155 May  1 06:09 ./data/osm_test_data/bad_neustadt_town.osm
-rw-r--r--  1 home  staff  1863096 May  1 06:09 ./tools/shaders_compiler/linux/mali_compiler_es3/openglessl/libMali-T600_r11p0-00rel0.so
-rw-r--r--  1 home  staff  1854526 May  1 06:09 ./android/app/src/fdroid/play/listings/en-US/graphics/phone-screenshots/1.jpg
-rw-r--r--  1 home  staff  1852120 May  1 06:09 ./tools/shaders_compiler/linux/mali_compiler_es3/openglessl/libMali-T600_r10p0-00rel0.so
```

I am trying to determine the most efficient approach for delving into the codebase to access the API that manages route optimization using map data.
